cosi/cosi.go

```diff
@@ -0,0 +1,502 @@
+/*
+Package cosi is the Collective Signing implementation according to the paper of
+Bryan Ford: http://arxiv.org/pdf/1503.08768v1.pdf .
+
+Stages of CoSi
+
+The CoSi-protocol has 4 stages:
+
+1. Announcement: The leader multicasts an announcement
+of the start of this round down through the spanning tree,
+optionally including the statement S to be signed.
+
+2. Commitment: Each node i picks a random secret vi and
+computes its individual commit Vi = Gvi . In a bottom-up
+process, each node i waits for an aggregate commit Vˆj from
+each immediate child j, if any. Node i then computes its
+own aggregate commit Vˆi = Vi \prod{j ∈ Cj}{Vˆj}, where Ci is the
+set of i’s immediate children. Finally, i passes Vi up to its
+parent, unless i is the leader (node 0).
+
+3. Challenge: The leader computes a collective challenge
+c = H( Aggregate Commit ∥ Aggregate Public key || Message ),
+then multicasts c down through the tree, along
+with the statement S to be signed if it was not already
+announced in phase 1.
+
+4. Response: In a final bottom-up phase, each node i waits
+to receive a partial aggregate response rˆj from each of
+its immediate children j ∈ Ci. Node i now computes its
+individual response ri = vi + cxi, and its partial aggregate
+response rˆi = ri + \sum{j ∈ Cj}{rˆj} . Node i finally passes rˆi
+up to its parent, unless i is the root.
+*/
+package cosi
+
+import (
+	cryptoRand "crypto/rand"
+	"crypto/sha512"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/dedis/crypto/abstract"
+	//own "github.com/nikkolasg/learning/crypto/util"
+)
+
+// Cosi is the struct that implements the basic cosi.
+type Cosi struct {
+	// Suite used
+	suite abstract.Suite
+	// publics is the list of public keys used for signing
+	publics []abstract.Point
+	// mask is the mask used to select which signers participated in this round
+	// or not. All code regarding the mask is directly inspired from
+	// github.com/bford/golang-x-crypto/ed25519/cosi code.
+	mask *CosiMask
+	// the longterm private key we use during the rounds
+	private abstract.Secret
+	// the message being co-signed
+	message []byte
+	// timestamp of when the announcement is done (i.e. timestamp of the four
+	// phases)
+	timestamp int64
+	// random is our own secret that we wish to commit during the commitment phase.
+	random abstract.Secret
+	// commitment is our own commitment
+	commitment abstract.Point
+	// V_hat is the aggregated commit (our own + the children's)
+	aggregateCommitment abstract.Point
+	// globalComitment is the global commit of all participants
+	//passed down in the challenge phase
+	globalCommitment abstract.Point
+	// challenge holds the challenge for this round
+	challenge abstract.Secret
+	// response is our own computed response
+	response abstract.Secret
+	// aggregateResponses is the aggregated response from the children + our own
+	aggregateResponse abstract.Secret
+}
```

Comment:
In the above, I think the fields should be organized so it's clear which fields are "public global information" - i.e., common information known to everyone, and which represent fields known only to a particular signer (e.g., private key, random, commitment).